core/src/stark/machine.rs

```diff
@@ -56,223 +41,46 @@ pub struct VerifyingKey<SC: StarkGenericConfig> {
-            self.memory_finalize.as_ref(),
-            self.program_memory_init.as_ref(),
-        ]
+    pub fn chips(&self) -> &[RiscvChip<SC>] {
+        &self.chips
     }
 
-    pub fn shard_chips(&self, shard: &ExecutionRecord) -> Vec<ChipRef<SC>> {
-        let mut chips = Vec::new();
-
-        chips.push(self.program.as_ref());
-        chips.push(self.cpu.as_ref());
-
-        if !shard.sha_extend_events.is_empty() {
-            chips.push(self.sha_extend.as_ref());
-        }
-
-        if !shard.sha_compress_events.is_empty() {
-            chips.push(self.sha_compress.as_ref());
-        }
-
-        if !shard.ed_add_events.is_empty() {
-            chips.push(self.ed_add_assign.as_ref());
-        }
-
-        if !shard.ed_decompress_events.is_empty() {
-            chips.push(self.ed_decompress.as_ref());
-        }
-
-        if !shard.k256_decompress_events.is_empty() {
-            chips.push(self.k256_decompress.as_ref());
-        }
-
-        if !shard.weierstrass_add_events.is_empty() {
-            chips.push(self.weierstrass_add_assign.as_ref());
-        }
-
-        if !shard.weierstrass_double_events.is_empty() {
-            chips.push(self.weierstrass_double_assign.as_ref());
-        }
-
-        if !shard.keccak_permute_events.is_empty() {
-            chips.push(self.keccak_permute.as_ref());
-        }
-
-        if !shard.blake3_compress_inner_events.is_empty() {
-            chips.push(self.blake3_compress_inner.as_ref());
-        }
-
-        if !shard.add_events.is_empty() {
-            chips.push(self.add.as_ref());
-        }
-
-        if !shard.sub_events.is_empty() {
-            chips.push(self.sub.as_ref());
-        }
-
-        if !shard.bitwise_events.is_empty() {
-            chips.push(self.bitwise.as_ref());
-        }
-
-        if !shard.divrem_events.is_empty() {
-            chips.push(self.div_rem.as_ref());
-        }
-
-        if !shard.mul_events.is_empty() {
-            chips.push(self.mul.as_ref());
-        }
-
-        if !shard.shift_right_events.is_empty() {
-            chips.push(self.shift_right.as_ref());
-        }
-
-        if !shard.shift_left_events.is_empty() {
-            chips.push(self.shift_left.as_ref());
-        }
-
-        if !shard.lt_events.is_empty() {
-            chips.push(self.lt.as_ref());
-        }
-
-        if !shard.field_events.is_empty() {
-            chips.push(self.field_ltu.as_ref());
-        }
-
-        if !shard.byte_lookups.is_empty() {
-            chips.push(self.byte.as_ref());
-        }
-
-        if !shard.first_memory_record.is_empty() {
-            chips.push(self.memory_init.as_ref());
-        }
-
-        if !shard.last_memory_record.is_empty() {
-            chips.push(self.memory_finalize.as_ref());
-        }
-
-        if !shard.program_memory_record.is_empty() {
-            chips.push(self.program_memory_init.as_ref());
-        }
+    pub fn dyn_chips(&self) -> Vec<ChipRef<SC>> {
```

Comment:
I feel like there def. should be a comment that this is used for non critical functions (e.g. debug_interactions).

Or maybe just do away with it?  Like go all in with derive?